### PR TITLE
MUL ID

### PR DESCRIPTION
**Pull request by SJuliez**
Adds the MUL ID to Entity and BLK/MTF files.
Also adds a hyperlink to our MechViews leading to the MUL page of that unit, when a MUL ID is available.
Best look at the commits separately as some contain several thousand unit files.

### CODE REVIEW

File: megamek/src/megamek/common/MechView.java [hunk 1]
```diff
@@ -1336,6 +1345,30 @@ public String toHTML() {
             return value + "<br/>\n";
         }
     }
+
+    /**
+     * Displays a hyperlink. Does not add a line break after itself.
+     */
+    private static class HyperLinkElement implements ViewElement {
+
+        private final String adress;
```
Comment (Windchild292) [inline megamek/src/megamek/common/MechView.java:1356]: I think you mean address 😉

File: megamek/src/megamek/common/MechView.java [hunk 2]
```diff
@@ -252,6 +252,8 @@ public MechView(final Entity entity, final boolean showDetail, final boolean use
                     Messages.getString("MechView.Source")));
             if (!entity.getSource().isEmpty()) {
                 sHead.add(new LabeledElement("", entity.getSource()));
+            } else {
+                sHead.add(new LabeledElement("", "(unkown)"));
```
Comment (Windchild292) [inline megamek/src/megamek/common/MechView.java:256]: It should be "unknown", not "unkown"

File: megamek/src/megamek/common/MechView.java [hunk 3]
```diff
@@ -245,11 +245,20 @@ public MechView(final Entity entity, final boolean showDetail, final boolean use
         if (useAlternateCost && entity.getAlternateCost() > 0) {
             cost = entity.getAlternateCost();
         }
-        sHead.add(new LabeledElement(Messages.getString("MechView.Cost"),//
+        sHead.add(new LabeledElement(Messages.getString("MechView.Cost"),
                 dFormatter.format(cost) + " C-bills"));
-        if (!entity.getSource().isEmpty()) {
-            sHead.add(new LabeledElement(Messages.getString("MechView.Source"), entity.getSource()));//
+        if (entity.hasMulId()) {
+            sHead.add(new HyperLinkElement("http://www.masterunitlist.info/Unit/Details/" + entity.getMulId(),
```
Comment (Windchild292) [inline megamek/src/megamek/common/MechView.java:251]: The actual link should probably be a MMConstant.

**Review by Windchild292 — COMMENTED** [review]
A pair of spelling mistakes and a suggestion for the location of a constant. Other than that, I'm good with the code for this PR.

File: megamek/src/megamek/common/MechView.java (hunk 1 above)
Comment (SJuliez) [reply megamek/src/megamek/common/MechView.java:1356]: indeed :)

File: megamek/src/megamek/common/MechView.java (hunk 2 above)
Comment (SJuliez) [reply megamek/src/megamek/common/MechView.java:256]: I must have been really tired and impatient...

File: megamek/src/megamek/common/MechSummary.java [hunk 4]
```diff
@@ -118,6 +119,8 @@ public String getModel() {
         return (m_sModel);
     }
 
+    public int getMulId() { return mulId; }
```
Comment (Windchild292) [inline megamek/src/megamek/common/MechSummary.java:122]: We don't do single line getters...

**Review by Windchild292 — COMMENTED** [review]
Just one style issue to fix

**Comment by HammerGS** [comment]
I've sent Juliez an excel sheet with another 450 odd matches to the MUL. Testing the branch I haven't seen any issues and think this should be considered for 49.7 depending on those new matches.

File: megamek/src/megamek/common/MechSummary.java (hunk 4 above)
Comment (SJuliez) [reply megamek/src/megamek/common/MechSummary.java:122]: lol this was an IntelliJ auto-add

**Comment by SJuliez** [comment]
Very nice! Of the 7016 units that can possibly have a MUL ID 6626 now have one, leaving 390 units without. 245 of those are Infantry that may not have a corresponding MUL entry. Of the remaining 145, 62 are Aero, 40 are Meks. So there's really only a handful of entries missing.

Concerning merging this, I'd prefer making informed decisions rather than following some rule. This PR changes thousands of unit files of which I've only really tested or looked at a few. It also changes the unit loading and saving code and makes the units not-backward-compatible (this is something that I could actually reverse, thinking about it now. I was doing what had been done in the past with other changes to mtf but it's not actually necessary). I'd consider this a very dangerous PR to merge because if it breaks something it's going to break it real hard.

**Comment by SJuliez** [comment]
Next round! Down to 226 of 7016 units (not unofficial, not Gun Emplacement) without MUL ID (27 Meks, 13 Aero, 167 Infantry, no PM, and 19 Tanks)

**Comment by HammerGS** [comment]
We've got unit files conflicting as I fixed them in master.

**Pull request merged**